Source/WTF/wtf/ThreadSafeWeakPtr.h

```diff
@@ -0,0 +1,195 @@
+ */
+
+#pragma once
+
+#include <wtf/Lock.h>
+#include <wtf/MainThread.h>
+#include <wtf/RefPtr.h>
+
+namespace WTF {
+
+template<typename> class ThreadSafeWeakPtr;
+template<typename, DestructionThread> class ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr;
+
+template<typename T>
+class ThreadSafeWeakPtrControlBlock {
+    WTF_MAKE_NONCOPYABLE(ThreadSafeWeakPtrControlBlock);
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
+public:
+    void ref() const
+    {
+        Locker locker { m_lock };
+        m_weakReferenceCount++;
+    }
+
+    void deref() const
+    {
+        bool shouldDeleteControlBlock { false };
+        {
+            Locker locker { m_lock };
+            RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_weakReferenceCount);
+            if (!--m_weakReferenceCount && !m_strongReferenceCount)
+                shouldDeleteControlBlock = true;
+        }
+        if (shouldDeleteControlBlock)
+            delete this;
+    }
+
+    void strongRef() const
+    {
+        Locker locker { m_lock };
+        RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_object);
+        m_strongReferenceCount++;
+    }
+
+    template<DestructionThread destructionThread>
+    void strongDeref() const
+    {
+        bool shouldDeleteControlBlock { false };
+        {
+            Locker locker { m_lock };
+            RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_object);
+            if (LIKELY(--m_strongReferenceCount))
+                return;
+
+            auto deleteObject = [object = std::exchange(m_object, nullptr)] {
+                delete static_cast<const T*>(object);
+            };
+            switch (destructionThread) {
+            case DestructionThread::Any:
+                deleteObject();
+                break;
+            case DestructionThread::Main:
+                ensureOnMainThread(WTFMove(deleteObject));
+                break;
+            case DestructionThread::MainRunLoop:
+                ensureOnMainRunLoop(WTFMove(deleteObject));
+                break;
+            }
+
+            if (!m_weakReferenceCount)
+                shouldDeleteControlBlock = true;
+        }
+        if (shouldDeleteControlBlock)
+            delete this;
+    }
+
+    RefPtr<T> makeStrongReferenceIfPossible() const
+    {
+        Locker locker { m_lock };
+        if (m_object) {
+            // Calling the RefPtr constructor would call strongRef() and deadlock.
+            m_strongReferenceCount++;
+            return adoptRef(m_object);
+        }
+        return nullptr;
+    }
+
+    bool objectHasBeenDeleted() const
+    {
+        Locker locker { m_lock };
+        return !m_object;
+    }
+
+private:
+    template<typename, DestructionThread> friend class ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr;
+    ThreadSafeWeakPtrControlBlock(T& object)
+        : m_object(&object) { }
+
+    mutable Lock m_lock;
+    mutable size_t m_strongReferenceCount WTF_GUARDED_BY_LOCK(m_lock) { 1 };
```

Comment:
I thought about that, but since either way it'll use the same amount of memory when padded (4 pointers size) I decided to be even more safe than ThreadSafeRefCounted in case someone makes 4 billion strong references.